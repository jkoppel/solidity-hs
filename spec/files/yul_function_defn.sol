contract Foo {
       function log(bytes32 p0) internal pure {
        bytes32 m0;
        bytes32 m1;
        bytes32 m2;
        bytes32 m3;
        assembly {
            function writeString(pos, w) {
                let length := 0
                for {} lt(length, 0x20) { length := add(length, 1) } { if iszero(byte(length, w)) { break } }
                mstore(pos, length)
                let shift := sub(256, shl(3, length))
                mstore(add(pos, 0x20), shl(shift, shr(shift, w)))
            }
            m0 := mload(0x00)
            m1 := mload(0x20)
            m2 := mload(0x40)
            m3 := mload(0x60)
            // Selector of `log(string)`.
            mstore(0x00, 0x41304fac)
            mstore(0x20, 0x20)
            writeString(0x40, p0)
        }
        _sendLogPayload(0x1c, 0x64);
        assembly {
            mstore(0x00, m0)
            mstore(0x20, m1)
            mstore(0x40, m2)
            mstore(0x60, m3)
        }
    }
}