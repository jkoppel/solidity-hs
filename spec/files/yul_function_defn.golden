[ Contract
    ( ContractDefinition
        { abstract = False
        , name = Identifier "Foo"
        , inheritance = Nothing
        , body =
            [ CFunction
                ( FunctionDefinition
                    { name = Identifier "log"
                    , params =
                        [ Parameter
                            { kind = ElementaryType
                                ( FixedBytes 32 )
                            , location = Nothing
                            , ident = Just
                                ( Identifier "p0" )
                            }
                        ]
                    , restrictions =
                        [ FuncVisibility FuncInternal
                        , FuncMutability Pure
                        ]
                    , returns = Nothing
                    , body = BlockStatement
                        [ VariableStatement
                            ( VariableDecStatement
                                ( VariableDeclaration
                                    { kind = ElementaryType
                                        ( FixedBytes 32 )
                                    , location = Nothing
                                    , ident = Identifier "m0"
                                    }
                                ) Nothing
                            )
                        , VariableStatement
                            ( VariableDecStatement
                                ( VariableDeclaration
                                    { kind = ElementaryType
                                        ( FixedBytes 32 )
                                    , location = Nothing
                                    , ident = Identifier "m1"
                                    }
                                ) Nothing
                            )
                        , VariableStatement
                            ( VariableDecStatement
                                ( VariableDeclaration
                                    { kind = ElementaryType
                                        ( FixedBytes 32 )
                                    , location = Nothing
                                    , ident = Identifier "m2"
                                    }
                                ) Nothing
                            )
                        , VariableStatement
                            ( VariableDecStatement
                                ( VariableDeclaration
                                    { kind = ElementaryType
                                        ( FixedBytes 32 )
                                    , location = Nothing
                                    , ident = Identifier "m3"
                                    }
                                ) Nothing
                            )
                        , Assembly
                            ( AssemblyStatement
                                { dialect = Nothing
                                , flags = []
                                , body =
                                    [ YulFunctionDefinition
                                        ( YulFunctionDefinitionDeclaration
                                            { ident = YulIdentifier
                                                ( Identifier "writeString" )
                                            , params =
                                                [ YulIdentifier
                                                    ( Identifier "pos" )
                                                , YulIdentifier
                                                    ( Identifier "w" )
                                                ]
                                            , returns = Nothing
                                            , body = YulBlock
                                                [ YulVariableStatement
                                                    ( YulVariableDeclaration
                                                        ( YulIdentifier
                                                            ( Identifier "length" )
                                                        )
                                                        ( YulExpressionLiteral
                                                            ( YulDecimalNumber 0 )
                                                        )
                                                    )
                                                , YulForStatement
                                                    ( YulForStatementDeclaration
                                                        { initializer = YulBlock []
                                                        , condition = YulExpressionFunctionCall
                                                            ( YulFunctionCallDeclaration
                                                                { ident = YulEvmBuiltin
                                                                    ( Identifier "lt" )
                                                                , body =
                                                                    [ YulExpressionPath
                                                                        ( YulIdentifierPath
                                                                            [ YulIdentifier
                                                                                ( Identifier "length" )
                                                                            ]
                                                                        )
                                                                    , YulExpressionLiteral
                                                                        ( YulHexNumber 32 )
                                                                    ]
                                                                }
                                                            )
                                                        , increment = YulBlock
                                                            [ YulAssignment
                                                                ( YulAssignmentDeclaration
                                                                    ( YulIdentifierPath
                                                                        [ YulIdentifier
                                                                            ( Identifier "length" )
                                                                        ]
                                                                    )
                                                                    ( YulExpressionFunctionCall
                                                                        ( YulFunctionCallDeclaration
                                                                            { ident = YulEvmBuiltin
                                                                                ( Identifier "add" )
                                                                            , body =
                                                                                [ YulExpressionPath
                                                                                    ( YulIdentifierPath
                                                                                        [ YulIdentifier
                                                                                            ( Identifier "length" )
                                                                                        ]
                                                                                    )
                                                                                , YulExpressionLiteral
                                                                                    ( YulDecimalNumber 1 )
                                                                                ]
                                                                            }
                                                                        )
                                                                    )
                                                                )
                                                            ]
                                                        , body = YulBlock
                                                            [ YulIfStatement
                                                                ( YulIfStatementDeclaration
                                                                    { expr = YulExpressionFunctionCall
                                                                        ( YulFunctionCallDeclaration
                                                                            { ident = YulEvmBuiltin
                                                                                ( Identifier "iszero" )
                                                                            , body =
                                                                                [ YulExpressionFunctionCall
                                                                                    ( YulFunctionCallDeclaration
                                                                                        { ident = YulEvmBuiltin
                                                                                            ( Identifier "byte" )
                                                                                        , body =
                                                                                            [ YulExpressionPath
                                                                                                ( YulIdentifierPath
                                                                                                    [ YulIdentifier
                                                                                                        ( Identifier "length" )
                                                                                                    ]
                                                                                                )
                                                                                            , YulExpressionPath
                                                                                                ( YulIdentifierPath
                                                                                                    [ YulIdentifier
                                                                                                        ( Identifier "w" )
                                                                                                    ]
                                                                                                )
                                                                                            ]
                                                                                        }
                                                                                    )
                                                                                ]
                                                                            }
                                                                        )
                                                                    , body = YulBlock [ YulBreak ]
                                                                    }
                                                                )
                                                            ]
                                                        }
                                                    )
                                                , YulFunctionCall
                                                    ( YulFunctionCallDeclaration
                                                        { ident = YulEvmBuiltin
                                                            ( Identifier "mstore" )
                                                        , body =
                                                            [ YulExpressionPath
                                                                ( YulIdentifierPath
                                                                    [ YulIdentifier
                                                                        ( Identifier "pos" )
                                                                    ]
                                                                )
                                                            , YulExpressionPath
                                                                ( YulIdentifierPath
                                                                    [ YulIdentifier
                                                                        ( Identifier "length" )
                                                                    ]
                                                                )
                                                            ]
                                                        }
                                                    )
                                                , YulVariableStatement
                                                    ( YulVariableDeclaration
                                                        ( YulIdentifier
                                                            ( Identifier "shift" )
                                                        )
                                                        ( YulExpressionFunctionCall
                                                            ( YulFunctionCallDeclaration
                                                                { ident = YulEvmBuiltin
                                                                    ( Identifier "sub" )
                                                                , body =
                                                                    [ YulExpressionLiteral
                                                                        ( YulDecimalNumber 256 )
                                                                    , YulExpressionFunctionCall
                                                                        ( YulFunctionCallDeclaration
                                                                            { ident = YulEvmBuiltin
                                                                                ( Identifier "shl" )
                                                                            , body =
                                                                                [ YulExpressionLiteral
                                                                                    ( YulDecimalNumber 3 )
                                                                                , YulExpressionPath
                                                                                    ( YulIdentifierPath
                                                                                        [ YulIdentifier
                                                                                            ( Identifier "length" )
                                                                                        ]
                                                                                    )
                                                                                ]
                                                                            }
                                                                        )
                                                                    ]
                                                                }
                                                            )
                                                        )
                                                    )
                                                , YulFunctionCall
                                                    ( YulFunctionCallDeclaration
                                                        { ident = YulEvmBuiltin
                                                            ( Identifier "mstore" )
                                                        , body =
                                                            [ YulExpressionFunctionCall
                                                                ( YulFunctionCallDeclaration
                                                                    { ident = YulEvmBuiltin
                                                                        ( Identifier "add" )
                                                                    , body =
                                                                        [ YulExpressionPath
                                                                            ( YulIdentifierPath
                                                                                [ YulIdentifier
                                                                                    ( Identifier "pos" )
                                                                                ]
                                                                            )
                                                                        , YulExpressionLiteral
                                                                            ( YulHexNumber 32 )
                                                                        ]
                                                                    }
                                                                )
                                                            , YulExpressionFunctionCall
                                                                ( YulFunctionCallDeclaration
                                                                    { ident = YulEvmBuiltin
                                                                        ( Identifier "shl" )
                                                                    , body =
                                                                        [ YulExpressionPath
                                                                            ( YulIdentifierPath
                                                                                [ YulIdentifier
                                                                                    ( Identifier "shift" )
                                                                                ]
                                                                            )
                                                                        , YulExpressionFunctionCall
                                                                            ( YulFunctionCallDeclaration
                                                                                { ident = YulEvmBuiltin
                                                                                    ( Identifier "shr" )
                                                                                , body =
                                                                                    [ YulExpressionPath
                                                                                        ( YulIdentifierPath
                                                                                            [ YulIdentifier
                                                                                                ( Identifier "shift" )
                                                                                            ]
                                                                                        )
                                                                                    , YulExpressionPath
                                                                                        ( YulIdentifierPath
                                                                                            [ YulIdentifier
                                                                                                ( Identifier "w" )
                                                                                            ]
                                                                                        )
                                                                                    ]
                                                                                }
                                                                            )
                                                                        ]
                                                                    }
                                                                )
                                                            ]
                                                        }
                                                    )
                                                ]
                                            }
                                        )
                                    , YulAssignment
                                        ( YulAssignmentDeclaration
                                            ( YulIdentifierPath
                                                [ YulIdentifier
                                                    ( Identifier "m0" )
                                                ]
                                            )
                                            ( YulExpressionFunctionCall
                                                ( YulFunctionCallDeclaration
                                                    { ident = YulEvmBuiltin
                                                        ( Identifier "mload" )
                                                    , body =
                                                        [ YulExpressionLiteral
                                                            ( YulHexNumber 0 )
                                                        ]
                                                    }
                                                )
                                            )
                                        )
                                    , YulAssignment
                                        ( YulAssignmentDeclaration
                                            ( YulIdentifierPath
                                                [ YulIdentifier
                                                    ( Identifier "m1" )
                                                ]
                                            )
                                            ( YulExpressionFunctionCall
                                                ( YulFunctionCallDeclaration
                                                    { ident = YulEvmBuiltin
                                                        ( Identifier "mload" )
                                                    , body =
                                                        [ YulExpressionLiteral
                                                            ( YulHexNumber 32 )
                                                        ]
                                                    }
                                                )
                                            )
                                        )
                                    , YulAssignment
                                        ( YulAssignmentDeclaration
                                            ( YulIdentifierPath
                                                [ YulIdentifier
                                                    ( Identifier "m2" )
                                                ]
                                            )
                                            ( YulExpressionFunctionCall
                                                ( YulFunctionCallDeclaration
                                                    { ident = YulEvmBuiltin
                                                        ( Identifier "mload" )
                                                    , body =
                                                        [ YulExpressionLiteral
                                                            ( YulHexNumber 64 )
                                                        ]
                                                    }
                                                )
                                            )
                                        )
                                    , YulAssignment
                                        ( YulAssignmentDeclaration
                                            ( YulIdentifierPath
                                                [ YulIdentifier
                                                    ( Identifier "m3" )
                                                ]
                                            )
                                            ( YulExpressionFunctionCall
                                                ( YulFunctionCallDeclaration
                                                    { ident = YulEvmBuiltin
                                                        ( Identifier "mload" )
                                                    , body =
                                                        [ YulExpressionLiteral
                                                            ( YulHexNumber 96 )
                                                        ]
                                                    }
                                                )
                                            )
                                        )
                                    , YulFunctionCall
                                        ( YulFunctionCallDeclaration
                                            { ident = YulEvmBuiltin
                                                ( Identifier "mstore" )
                                            , body =
                                                [ YulExpressionLiteral
                                                    ( YulHexNumber 0 )
                                                , YulExpressionLiteral
                                                    ( YulHexNumber 1093685164 )
                                                ]
                                            }
                                        )
                                    , YulFunctionCall
                                        ( YulFunctionCallDeclaration
                                            { ident = YulEvmBuiltin
                                                ( Identifier "mstore" )
                                            , body =
                                                [ YulExpressionLiteral
                                                    ( YulHexNumber 32 )
                                                , YulExpressionLiteral
                                                    ( YulHexNumber 32 )
                                                ]
                                            }
                                        )
                                    , YulFunctionCall
                                        ( YulFunctionCallDeclaration
                                            { ident = YulIdentifier
                                                ( Identifier "writeString" )
                                            , body =
                                                [ YulExpressionLiteral
                                                    ( YulHexNumber 64 )
                                                , YulExpressionPath
                                                    ( YulIdentifierPath
                                                        [ YulIdentifier
                                                            ( Identifier "p0" )
                                                        ]
                                                    )
                                                ]
                                            }
                                        )
                                    ]
                                }
                            )
                        , ExpressionStatement
                            ( FunctionCall
                                ( IdentifierExpression
                                    ( Identifier "_sendLogPayload" )
                                )
                                ( CommaArguments
                                    [ ExpressionLiteral
                                        ( NumberLiteral 28.0 Nothing )
                                    , ExpressionLiteral
                                        ( NumberLiteral 100.0 Nothing )
                                    ]
                                )
                            )
                        , Assembly
                            ( AssemblyStatement
                                { dialect = Nothing
                                , flags = []
                                , body =
                                    [ YulFunctionCall
                                        ( YulFunctionCallDeclaration
                                            { ident = YulEvmBuiltin
                                                ( Identifier "mstore" )
                                            , body =
                                                [ YulExpressionLiteral
                                                    ( YulHexNumber 0 )
                                                , YulExpressionPath
                                                    ( YulIdentifierPath
                                                        [ YulIdentifier
                                                            ( Identifier "m0" )
                                                        ]
                                                    )
                                                ]
                                            }
                                        )
                                    , YulFunctionCall
                                        ( YulFunctionCallDeclaration
                                            { ident = YulEvmBuiltin
                                                ( Identifier "mstore" )
                                            , body =
                                                [ YulExpressionLiteral
                                                    ( YulHexNumber 32 )
                                                , YulExpressionPath
                                                    ( YulIdentifierPath
                                                        [ YulIdentifier
                                                            ( Identifier "m1" )
                                                        ]
                                                    )
                                                ]
                                            }
                                        )
                                    , YulFunctionCall
                                        ( YulFunctionCallDeclaration
                                            { ident = YulEvmBuiltin
                                                ( Identifier "mstore" )
                                            , body =
                                                [ YulExpressionLiteral
                                                    ( YulHexNumber 64 )
                                                , YulExpressionPath
                                                    ( YulIdentifierPath
                                                        [ YulIdentifier
                                                            ( Identifier "m2" )
                                                        ]
                                                    )
                                                ]
                                            }
                                        )
                                    , YulFunctionCall
                                        ( YulFunctionCallDeclaration
                                            { ident = YulEvmBuiltin
                                                ( Identifier "mstore" )
                                            , body =
                                                [ YulExpressionLiteral
                                                    ( YulHexNumber 96 )
                                                , YulExpressionPath
                                                    ( YulIdentifierPath
                                                        [ YulIdentifier
                                                            ( Identifier "m3" )
                                                        ]
                                                    )
                                                ]
                                            }
                                        )
                                    ]
                                }
                            )
                        ]
                    }
                )
            ]
        }
    )
]